src/artemis/fgs_communicator.py

```diff
@@ -0,0 +1,114 @@
+import os
+import time
+
+from bluesky.callbacks import CallbackBase
+
+import artemis.log
+from artemis.ispyb.store_in_ispyb import StoreInIspyb2D, StoreInIspyb3D
+from artemis.nexus_writing.write_nexus import (
+    NexusWriter,
+    create_parameters_for_first_file,
+    create_parameters_for_second_file,
+)
+from artemis.parameters import FullParameters
+from artemis.zocalo_interaction import run_end, run_start, wait_for_result
+
+
+class FGSCommunicator(CallbackBase):
+    """Class for external communication (e.g. ispyb, zocalo...) during Artemis
+    grid scan experiments.
+
+    Listens to documents emitted by the RE and:
+    - prepares nexus files
+    - prepares ipsyb deposition
+    - submits job to zocalo
+    """
+
+    def __init__(self):
+        self.reset(FullParameters(), write_files=False)
+
+    def reset(self, parameters: FullParameters, write_files=True):
+        self.params = parameters
+        ispyb_config = os.environ.get("ISPYB_CONFIG_PATH", "TEST_CONFIG")
+        self.ispyb = (
+            StoreInIspyb3D(ispyb_config, self.params)
+            if self.params.grid_scan_params.is_3d_grid_scan
+            else StoreInIspyb2D(ispyb_config, self.params)
+        )
+        self.processing_start_time = 0.0
+        self.processing_time = 0.0
+        if write_files:
+            self.nxs_writer_1 = NexusWriter(
+                create_parameters_for_first_file(self.params)
+            )
+            self.nxs_writer_2 = NexusWriter(
+                create_parameters_for_second_file(self.params)
+            )
+        self.results = None
+        self.xray_centre_motor_position = None
+        self.ispyb_ids: tuple = (None, None, None)
+        self.datacollection_group_id = None
+
+    def start(self, doc: dict):
+        artemis.log.LOGGER.debug(f"\n\nReceived start document:\n\n {doc}\n")
+        artemis.log.LOGGER.info("Creating Nexus files.")
+        self.nxs_writer_1.create_nexus_file()
+        self.nxs_writer_2.create_nexus_file()
+
+    def event(self, doc: dict):
+        artemis.log.LOGGER.debug(f"\n\nReceived event document:\n{doc}\n")
+
+        if doc.get("name") == "ispyb_motor_positions":
```

Comment:
Should: I don't like the hardcoded string here, it means if we change the name of the doc in the plan we have to change it here or else it will break. I think it would be better if they were both looking at a constant stored elsewhere so we could ensure their always the same.

Reply:
ok, i've made them read from a yml config file which we can put plan names in, but actually do you think it's better to just have them in  a python file instead?

Reply:
I think a python file with a constant in is better. With the YAML file we have a similar issue where both files need to know the name of the key. If it's a real constant you can change the name by doing a refactor in the IDE much more easily